### ISSUE TITLE
Unable to install grpc on Ubuntu 16.04

### ISSUE THREAD

**Issue by daemonslayer** [grpc/grpc#9549]
After installing protobuf, I used these commands to build grpc:
```
$ git clone https://github.com/grpc/grpc.git
$ cd grpc
$ git submodule update --init
$ make  # <-- Here's where things exploded
$ make install
```
I am getting this error: 
```
[MAKE]    Generating cache.mk
[HOSTLD]  Linking /home/cortana/Libraries/grpc/bins/opt/grpc_cpp_plugin
[HOSTLD]  Linking /home/cortana/Libraries/grpc/bins/opt/grpc_csharp_plugin
[HOSTLD]  Linking /home/cortana/Libraries/grpc/bins/opt/grpc_node_plugin
/usr/bin/ld: warning: libprotobuf.so.10, needed by //usr/local/lib/libprotoc.so, may conflict with libprotobuf.so.9
/usr/bin/ld: /home/cortana/Libraries/grpc/libs/opt/libgrpc_plugin_support.a(csharp_generator.o): undefined reference to symbol '_ZN6google8protobuf2io7Printer5PrintEPKcS4_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES4_SC_S4_SC_S4_SC_'
/usr/local/lib/libprotobuf.so.10: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
Makefile:11195: recipe for target '/home/cortana/Libraries/grpc/bins/opt/grpc_csharp_plugin' failed
make: *** [/home/cortana/Libraries/grpc/bins/opt/grpc_csharp_plugin] Error 1
make: *** Waiting for unfinished jobs....
/usr/bin/ld: warning: libprotobuf.so.10, needed by //usr/local/lib/libprotoc.so, may conflict with libprotobuf.so.9
/usr/bin/ld: warning: libprotobuf.so.10, needed by //usr/local/lib/libprotoc.so, may conflict with libprotobuf.so.9
```
I had previously installed protobuf 3.2(latest) from release package but that wasn't supported. So I `make uninstall`ed it and installed protobuf 3.0.0. On doing `protoc --version`, I get `libprotoc 3.0.0`.

**Comment by valorcurse** [comment]
I had the same problem on Ubuntu 16 while trying to build gRPC using Protobuf 3.2.0. This happens because the Makefile uses the system libprotobuf libraries while building the plugins, instead of using the protobuf library installed in /usr/local/lib. 

I managed to build the library by modifying the _HOST_LDLIBS_PROTOC_ variable in the Makefile from: 

`HOST_LDLIBS_PROTOC += $(addprefix -l, $(LIBS_PROTOC))
`

to:

`HOST_LDLIBS_PROTOC += -L/usr/local/lib $(addprefix -l, $(LIBS_PROTOC))
`

Do a  `make clean` and `make` again.

I couldn't find a cleaner way of doing this as of yet, but you should be able to get it working.

**Comment by chakpongchung** [comment]
[HOSTLD]  Linking /data/home/microway/Software/grpc/bins/opt/grpc_ruby_plugin
[HOSTLD]  Linking /data/home/microway/Software/grpc/bins/opt/grpc_node_plugin
/usr/bin/ld: /data/home/microway/Software/grpc/libs/opt/libgrpc_plugin_support.a(csharp_generator.o): undefined reference to symbol '_ZN6google8protobuf2io7Printer5PrintEPKcS4_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES4_SC_S4_SC_S4_SC_'
//usr/local/lib/libprotobuf.so.12: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
Makefile:14053: recipe for target '/data/home/microway/Software/grpc/bins/opt/grpc_csharp_plugin' failed



@valorcurse  your solution does not work on my computer.


On doing protoc --version, I get libprotoc 3.2.0

**Comment by nicolasnoble** [comment]
In general, this means you have several versions of protobuf / protoc installed, and they unfortunately don't play well with each other.

**Comment by mehrdada** [comment]
I'm closing this issue. Please do re-open the issue with information on how to reproduce if you experience it with only one supported version of protobuf installed.

**Comment by jayeye** [comment]
@mehrdada not being able to play well with multiple versions of the protobuf library is a bug. That's why libraries are versioned, so that different packages can be built with different verions. grpc explicitly imports protobuf as a submodule; it should build against *that* and ignore any existing versions.

**Comment by mehrdada** [comment]
@jayeye I do relate with your sentiment. Would you mind filing another, more targeted and specific, issue titled as such (building fine despite other versions of protoc installed) and to track this exact concern (ideally with a new report of under what conditions this is reproducible using the latest release or master of gRPC, as some past issues may have been fixed already)?

**Comment by jhollowayj** [comment]
@jayeye, @mehrdada, Was there a new issue added to track that new issue?

**Comment by vladkras** [comment]
as mentioned in https://github.com/grpc/grpc/issues/4105#issuecomment-248094747 I had to do
```
$ git clone https://github.com/grpc/grpc.git
$ cd grpc
$ git submodule update --init
```
as written in docs but then
```
 $  cd third_party/protobuf
 $  ./autogen.sh
 $  cd ../..
```
and only after that
```
$ make
```
started to work (in my case [make grpc_php_plugin](https://grpc.io/docs/tutorials/basic/php.html#generating-client-code)). 

libtool and other packages were installed before with
```
sudo apt-get install build-essential libtool autoconf automake
```
so just installing them is required to avoid different errors like `configure.ac:104: error: possibly undefined macro: AC_PROG_LIBTOOL` and `autoreconf: not found` but is not a complete solution

**Comment by alishir** [comment]
I have the same issue. Before installing grpc I have been installed to version of protobuf, one version via `apt` and other from source, I tried to remove `apt` installed version but many packages depends on that, so I didn't uninstall the `apt` version. Then I uninstalled the version that installed from source and grpc installed successfully.
